### PR TITLE
Full mock coverage

### LINKED ISSUE #12: automock functions not yet mockable
this is follow up to #6 and #7

### PR DESCRIPTION
**Pull request by WFinck97**
an hdk_crud refactor to be able to mock all the exposed functions for unit testing on the consumer side.

closes #12

- [x] convert all `inner_` crud functions into generic `actions`
- [x] update comments to describe the generic actions
- [x] use mock feature
- [x] make the remaining hdk_crud functions mockable (turn into struct methods)

Follow ups (split into separate issue and PR)
- [ ] write unit tests for remaining functions: #1 
- [ ] fetchers constructor: #11

### CODE REVIEW

**Comment by WFinck97** [comment]
### make remaining hdk_functions mockable
current plan:
1. split remaining functions in `retrieval.rs` into their own file and make them mockable (like [get_latest_for_entry.rs](https://github.com/lightningrodlabs/hdk_crud/blob/main/src/retrieval/get_latest_for_entry.rs))
image

2. make generic crud action functions (in crud.rs) mockable. 
- question: each function will become a struct method such that it can be mocked. Should these structs be defined in crud.rs, or split into their own files like the other mocked functions?

**Comment by Connoropolous** [comment]
@WFinck97 
I'm thinking... 
along with 'retrieval' which already exists, there could be `chain_actions` mod... 
in it would be a file for `create_action` `update_action` `delete_action`. Separated because these actions affect the source chain

File: src/chain_actions/fetch_action.rs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        path: Path,
+    ) -> ExternResult<Vec<WireElement<T>>>
+    where
+        Entry: TryFrom<T, Error = E>,
+        WasmError: From<E>,
+        T: 'static + Clone + TryFrom<SerializedBytes, Error = SerializedBytesError>,
+        E: 'static,
+    {
+        let get_latest = crate::retrieval::get_latest_for_entry::GetLatestEntry {};
+        let entries = fetch_entries.fetch_entries::<T>( // TODO: change to struct method
```
Comment (Connoropolous) [inline src/chain_actions/fetch_action.rs:27]:
```suggestion
        let entries = fetch_entries.fetch_entries::<T>(
```

File: src/retrieval/utils.rs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+use hdk::prelude::*;
+
+/// A triple of an Entry along with the HeaderHash
+/// of that committed entry and the EntryHash of the entry
+pub type EntryAndHash<T> = (T, HeaderHash, EntryHash);
+
+/// The same as an EntryAndHash but inside an Option,
+/// so it can be Some(...) or None
+pub type OptionEntryAndHash<T> = Option<EntryAndHash<T>>;
```
Comment (Connoropolous) [inline src/retrieval/utils.rs:9]: would prefer these to be in a spot like 'types' or something, but not going to ask that to be done right now.

File: src/crud.rs [hunk 3]
```diff
@@ -1,14 +1,12 @@
...
 /// to having a full create-read-update-delete set of
-/// functionality in your Zome, plus "signals" (events).
+/// functionality in your Zome, "signals" (events), as well as
+/// time based queries.
 /// See [example] for a comprehensive look at how this works.
 /// ```ignore
 /// use hdk::prelude::*;
 /// use hdk_crud::*;
-///
 /// #[hdk_entry(id = "test")]
```
Comment (Connoropolous) [inline src/crud.rs:10]:
```suggestion
/// #[hdk_entry(id = "example")]
```

File: src/crud.rs [hunk 4]
```diff
@@ -1,14 +1,12 @@
...
 /// #[derive(Clone, PartialEq)]
 /// pub struct Test {
```
Comment (Connoropolous) [inline src/crud.rs:12]:
```suggestion
/// pub struct Example {
```

File: src/chain_actions/create_action.rs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+use holo_hash::{AgentPubKey, EntryHashB64, HeaderHashB64};
+use crate::wire_element::WireElement;
+use crate::chain_actions::utils::now_date_time;
+
+#[cfg(feature = "mock")]
+use ::mockall::automock;
+
+trait ConvertReceiverSignal<T, S> {
+    fn convert_to_receiver_signal(action_signal: crate::signals::ActionSignal<T>) -> S;
+}
```
Comment (Connoropolous) [inline src/chain_actions/create_action.rs:12]:
```suggestion
```

**Review by Connoropolous — CHANGES_REQUESTED** [review]
I think all my change requests are expressed as mergeable suggestions. 
You can use the bulk 'add suggestion to batch' feature to merge them all at once. 

So close! Excellent

**Pull request merged**